### ISSUE TITLE
:tag does not jump back to the correct nr tag

### ISSUE THREAD

**Issue by andymass** [vim/vim#3944]
In the example below, typing `2<c-]>` on `func1` jumps correctly:

```
  # TO tag         FROM line  in file/text
  1  2 func1               6  int func1()
>
```

but typing `<c-t>` then `:tag` does not return to nr 2, and instead replaces tagstack with nr 1.
```
  # TO tag         FROM line  in file/text
  1  1 func1               6  int func1()
>
```

I believe this problem was introduced by patch 7.4.746 (https://github.com/vim/vim/commit/01cf376da1726862afc8fa1d84cf5a773909fd0d).

ex.c:
```c
struct foo
{
	int func1;
};
int func1()
{
}
```

tags:
```
!_TAG_FILE_FORMAT	2	/extended format; --format=1 will not append ;" to lines/
!_TAG_FILE_SORTED	1	/0=unsorted, 1=sorted, 2=foldcase/
!_TAG_OUTPUT_MODE	u-ctags	/u-ctags or e-ctags/
!_TAG_PROGRAM_AUTHOR	Universal Ctags Team	//
!_TAG_PROGRAM_NAME	Universal Ctags	/Derived from Exuberant Ctags/
!_TAG_PROGRAM_URL	https://ctags.io/	/official site/
!_TAG_PROGRAM_VERSION	0.0.0	/3d01d993/
foo	ex.c	/^struct foo$/;"	s	file:
func1	ex.c	/^	int func1;$/;"	m	struct:foo	typeref:typename:int	file:
func1	ex.c	/^int func1()$/;"	f	typeref:typename:int
```

**Comment by EgZvor** [comment]
It looks like this could be enough to fix this. I will use this build for a while and try to see if there will be any problems.
```
diff --git a/src/tag.c b/src/tag.c
index b1915e1e1..685f3d4c1 100644
--- a/src/tag.c
+++ b/src/tag.c
@@ -583,7 +583,7 @@ do_tag(
 	    }
 	    else
 #endif
-	    if (type == DT_TAG)
+	    if (type == DT_TAG && *tag != NUL)
 		/*
 		 * If a count is supplied to the ":tag <name>" command, then
 		 * jump to count'th matching tag.
```

**Comment by andymass** [comment]
Erop said:

> It looks like this could be enough to fix this. I will use this build for a while and try to see if there will be any problems.
>
> ```patch
> diff --git a/src/tag.c b/src/tag.c
> index b1915e1e1..685f3d4c1 100644
> --- a/src/tag.c
> +++ b/src/tag.c
> @@ -583,7 +583,7 @@ do_tag(
>  	    }
>  	    else
>  #endif
> -	    if (type == DT_TAG)
> +	    if (type == DT_TAG && *tag != NUL)
>  		/*
>  		 * If a count is supplied to the ":tag <name>" command, then
>  		 * jump to count'th matching tag.
> ```

Could you add a test for this, and make sure there is a test to prevent regression of 7.4.746? 
 There is `test_tagjump.vim` but I'm not sure what it covers.

**Comment by EgZvor** [comment]
Yeah, I removed the comment, because it does not work). It jumps to the last tag instead of the first with this and possibly breaks something else.

**Comment by EgZvor** [comment]
I dug a little bit deeper into this issue and I think there was never a correct behaviour in place. There is an inherent issue with `:tag`. When we do not provide the count explicitly it is assumed to be 1, but with this command we need to somehow differentiate between the explicit and implicit count equal to 1. In 7.4.745 `:1tag` still jumps to the last selected match and not to the `count`th match.

**Comment by andymass** [comment]
Well, `:{count}tag` and `:{count}tag {arg}` are really different commands. `:1tag` == `:tag` means go forward one in the tag stack.  `:1tag name` == `:tag name`  means go to the first matching tag for name (precise order is undocumented).

I agree the behavior was never really correct (before or after 7.4.745), but I think the implicit count is OK in this case.

**Comment by EgZvor** [comment]
Thank you for clarification, I forgot about `count` as a number of jumps in the existing stack. Here's what I come up with so far
```
diff --git a/src/tag.c b/src/tag.c
index b1915e1e1..bcdc8f832 100644
--- a/src/tag.c
+++ b/src/tag.c
@@ -504,13 +504,16 @@ do_tag(
 		tagmatchname = vim_strsave(name);
 	    }
 
-	    if (type == DT_TAG || type == DT_SELECT || type == DT_JUMP
+	    if (type == DT_SELECT || type == DT_JUMP
 #if defined(FEAT_QUICKFIX)
 		|| type == DT_LTAG
 #endif
 		)
 		cur_match = MAXCOL - 1;
-	    max_num_matches = cur_match + 1;
+	    if (type == DT_TAG)
+		max_num_matches = MAXCOL;
+	    else
+		max_num_matches = cur_match + 1;
 
 	    /* when the argument starts with '/', use it as a regexp */
 	    if (!no_regexp && *name == '/')
@@ -583,7 +586,7 @@ do_tag(
 	    }
 	    else
 #endif
-	    if (type == DT_TAG)
+	    if (type == DT_TAG && *tag != NUL)
 		/*
 		 * If a count is supplied to the ":tag <name>" command, then
 		 * jump to count'th matching tag.
```

**Comment by andymass** [comment]
Your patch appears to have passed a basic test, without regressing 7.4.746: https://github.com/vim/vim/pull/4177

Can you briefly explain the steps taken for the proposed fix?  It's very hard to follow the existing tag code, e.g. MAXCOL is confusing.  Also if you have any ideas for test cases it might be good to add to them to the PR.

**Comment by EgZvor** [comment]
The `MAXCOL` change is a fix of the regression introduced in 7.4.746. I diffed 7.4.745 against master and tried to trace why the `cur_match` gets zeroed. The regression was that `max_num_matches` was updated based on `cur_match`, but in case of newly added `type == DT_TAG` we don't want to update `cur_match`, yet want to update `max_num_matches`. That's why I removed `type == DT_TAG` from the first condition and added another one. I'm sorry it that's poorly worded. The second part is just syncing the code with the comment.

I'll try to come up with some test cases on a weekend. I once had a really strange behaviour of current match being correctly processed up to number 23 and for 24, 26 it was still saved as 23. Unfortunately I wasn't able to reproduce it again, so maybe it was some coding / compiling mistake I made.

**Comment by brammool** [comment]
> The `MAXCOL` change is a fix of the regression introduced in 7.4.746.
> I diffed 7.4.745 against master and tried to trace why the `cur_match`
> gets zeroed. The regression was that `max_num_matches` was updated
> based on `cur_match`, but in case of newly added `type == DT_TAG` we
> don't want to update `cur_match`, yet want to update
> `max_num_matches`. That's why I removed `type == DT_TAG` from the
> first condition and added another one. I'm sorry it that's poorly
> worded. The second part is just syncing the code with the comment.
> 
> I'll try to come up with some test cases on a weekend. I once had a
> really strange behaviour of current match being correctly processed up
> to number 23 and for 24, 26 it was still saved as 23. Unfortunately I
> wasn't able to reproduce it again, so maybe it was some coding /
> compiling mistake I made.

Pull request 4177 claims to have a test for this.
Please check it out.

-- 
hundred-and-one symptoms of being an internet addict:
143. You dream in pallettes of 216 websafe colors.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by h-east** [comment]
@andymass Thanks for fixing the bug I embedded in 7.4.746 :+1: (About 4 years ago!)

--
Best regards,
Hirohito Higashi (h_east)

**Comment by andymass** [comment]
@h-east: @EgZvor does deserve most of the credit :)   While you're here, is there any chance you can check over the patch?
